lib/chef/knife/cloud/google_service.rb

```diff
@@ -0,0 +1,503 @@
+
+      ui.msg("Instance #{instance_name} deleted successfully.")
+    end
+
+    def get_server(instance_name)
+      connection.get_instance(project, zone, instance_name)
+    end
+
+    def create_disk(name, size, type, source_image = nil)
+      disk = Google::Apis::ComputeV1::Disk.new
+      disk.name    = name
+      disk.size_gb = size
+      disk.type    = disk_type_url_for(type)
+
+      ui.msg("Creating a #{size} GB disk named #{name}...")
+
+      wait_for_operation(connection.insert_disk(project, zone, disk, source_image: source_image))
+
+      ui.msg("Waiting for disk to be ready...")
+
+      wait_for_status("READY") { connection.get_disk(project, zone, name) }
+
+      ui.msg("Disk created successfully.")
+    end
+
+    def delete_disk(name)
+      begin
+        connection.get_disk(project, zone, name)
+      rescue Google::Apis::ClientError
+        ui.warn("Unable to locate disk #{name} in project #{project}, zone #{zone}")
+        return
+      end
+
+      ui.confirm("Do you really want to delete disk #{name}")
+
+      ui.msg("Deleting disk #{name}...")
+      wait_for_operation(connection.delete_disk(project, zone, name))
+      ui.msg("Disk #{name} deleted successfully.")
+    end
+
+    def list_servers
+      instances = paginated_results(:list_instances, :items, project, zone)
+      return [] if instances.nil?
+
+      instances.each_with_object([]) do |instance, memo|
+        memo << OpenStruct.new(
+          name:         instance.name,
+          status:       instance.status,
+          machine_type: machine_type_for(instance),
+          network:      network_for(instance),
+          private_ip:   private_ip_for(instance),
+          public_ip:    public_ip_for(instance)
+        )
+      end
+    end
+
+    def list_zones
+      zones = paginated_results(:list_zones, :items, project)
+      return [] if zones.nil?
+
+      zones
+    end
+
+    def list_disks
+      disks = paginated_results(:list_disks, :items, project, zone)
+      return [] if disks.nil?
+
+      disks
+    end
+
+    def list_regions
+      regions = paginated_results(:list_regions, :items, project)
+      return [] if regions.nil?
+
+      regions
+    end
+
+    def list_project_quotas
+      quotas = connection.get_project(project).quotas
+      return [] if quotas.nil?
+
+      quotas
+    end
+
+    def validate_server_create_options!(options)
+      raise "Invalid machine type: #{options[:machine_type]}" unless valid_machine_type?(options[:machine_type])
+      raise "Invalid network: #{options[:network]}" unless valid_network?(options[:network])
+      raise "Invalid Public IP setting: #{options[:public_ip]}" unless valid_public_ip_setting?(options[:public_ip])
+      raise "Invalid image: #{options[:image]} - check your image name, or set an image project if needed" if image_search_for(options[:image], options[:image_project]).nil?
+    end
+
+    def check_api_call(&block)
```

Comment:
Curious as to why you validate with multiple extra API calls instead of just trying to handing the errors.  Seems less efficient but perhaps that makes it more user friendly?